In [1]:
#TD for to dos

# for timeline analysis : prophet

# Part 1: Cleaning and Engineering

<html>
   <head>
      <title>Table part1</title>
   </head>
   <body>
      <h1><font size="4">Table of content</font></h1>
      <ul style="list-style-type:none;line-height: 2;">
         <li><font size="3"><a href='#1'>1. Loading the Data</a></font></li>
         <li><font size="3"><a href='#2'>2. First Look at the Data</a></font></li>
         <ul style="margin-left: 1em; padding-left: 1em;list-style-type:none">
         <li><font size="3"><a href='#2.1'>2.1 Feature Dictionary</a></font></li>
         </ul>
         <li><font size="3"><a href='#3'>3. Data Cleaning</a></font></li>
         <ul style="margin-left: 1em; padding-left: 1em;list-style-type:none">
         <li><font size="3"><a href='#3.1'>3.1 Correcting Visit ID</a></font></li>
         <li><font size="3"><a href='#3.2'>3.2 Correcting Product Group and Line</a></font></li>
         <li><font size="3"><a href='#3.3'>3.3 NaNs</a></font></li>
         <li><font size="3"><a href='#3.4'>3.4 Data Types</a></font></li>
         <li><font size="3"><a href='#3.5'>3.5 Duplicates</a></font></li>         
         </ul>          
         <li><font size="3"><a href='#4'>4. Understanding particular Features in Detail</a></font></li>
         <li><font size="3"><a href='#5'>5. Feature Engineering</a></font></li>
         <ul style="margin-left: 1em; padding-left: 1em;list-style-type:none">
         <li><font size="3"><a href='#3.1'>5.1 Cities by Code</a></font></li>
         <li><font size="3"><a href='#3.2'>5.2 Timespan</a></font></li>
         <li><font size="3"><a href='#3.3'>5.3 Weekdays</a></font></li>
         <li><font size="3"><a href='#3.4'>5.4 Article Genres</a></font></li>
         <li><font size="3"><a href='#3.5'>5.5 Total price per article purchase</a></font></li>    
         <li><font size="3"><a href='#3.2'>5.6 Value and number of items of shopping carts</a></font></li>
         <li><font size="3"><a href='#3.3'>5.7 Stockprice per day</a></font></li>
         <li><font size="3"><a href='#3.4'>5.8 First date of purchase and purchase type</a></font></li>
         <li><font size="3"><a href='#3.5'>5.9 Final Feature Dictionary</a></font></li>               
         </ul>            
      </ul>
   </body>
</html>

## Importing libraries

In [2]:
# Importing the necessary libraries
import re
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import glob
import datetime
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import roc_auc_score
from sklearn.metrics import roc_curve, auc
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.decomposition import PCA
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import confusion_matrix, f1_score, precision_recall_fscore_support
from sklearn import preprocessing
from sklearn.model_selection import cross_val_score, cross_val_predict

## Setting options

In [3]:
pd.set_option('display.max_columns', 100)
pd.set_option('display.max_rows', 400)

In [4]:
#adjusting notebook width
#from IPython.core.display import display, HTML
#display(HTML("<style>.container { width:90% !important; }</style>"))

In [5]:
#markdowntables layout

In [6]:
%%html
<style>
table {float:left}
table td, table th, table tr {text-align:left !important;}
</style>

<a id='1'></a>

# 1. Loading the data

In [7]:
#sales dataframe
df_base = pd.read_csv("./Data/TTdata.csv")

//anaconda3/envs/nf2/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (5,20) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [8]:
#copy for working
df = df_base.copy()

In [9]:
#dataframe with citynames
#https://developers.google.com/adwords/api/docs/appendix/geotargeting
df_geotargets = pd.read_csv("./Data/geotargets.csv") 

In [10]:
df_stock = pd.read_csv("./Data/daily_TTI.DE.csv")

<a id='2'></a>

# 2. First Look at the Data

In [11]:
df.head()

visit_id_n  unser_id_n  visit_number        date  \
0      551960      553811           1.0  20180910.0   
1      242735      254033           1.0  20180104.0   
2      680200      620544           2.0  20180612.0   
3       44920      609593          46.0  20180913.0   
4      704159       73553           2.0  20180729.0   

                article_name    article_id               ga_category  \
0  Geschenk Gutschein 10 EUR  200000000010                    Herren   
1   Geschenkgutschein 10 EUR  200000000010  Damen/Geschenkgutscheine   
2  Geschenk Gutschein 10 EUR  200000000010  Damen/Geschenkgutscheine   
3  Geschenk Gutschein 10 EUR  200000000010                    Herren   
4  Geschenk Gutschein 10 EUR  200000000010                    Herren   

       brand  price  quantity device_category device_browser device_os  \
0  (not set)   10.0       1.0         desktop        Firefox   Windows   
1  (not set)   10.0       1.0         desktop        Firefox   Windows   
2  (not set)   10.0       1.0         desktop         Chrome   Windows   
3  (not set)   10.0       1.0          mobile         Safari       iOS   
4  (not set)   10.0       1.0         desktop         Chrome   Windows   

   country        region  city_id product_line product_group product_line_r  \
0  Germany  Lower Saxony  1004478       Herren        Herren         Herren   
1  Germany  Lower Saxony  9048892       Herren        Herren         Herren   
2  Germany        Berlin  1003854       Herren        Herren         Herren   
3  Germany  Lower Saxony  1004564       Herren        Herren         Herren   
4  Germany       Bavaria  1004246       Herren        Herren         Herren   

  product_group_r campaign_involved  
0          Herren             False  
1          Herren              True  
2          Herren              True  
3          Herren              True  
4          Herren              True

In [12]:
df.tail()

visit_id_n  unser_id_n  visit_number        date  \
3016738       33526      316730           2.0  20181122.0   
3016739      520195       51192          16.0  20181201.0   
3016740       18132      393215           2.0  20181124.0   
3016741        7155      211171           NaN         NaN   
3016742        7155      211171           NaN         NaN   

                  article_name          article_id             ga_category  \
3016738  Unifarbener Ösenschal       4260478143821               (not set)   
3016739  Unifarbener Ösenschal       4260478143821  Home & Living/Vorhänge   
3016740  Unifarbener Ösenschal       4260478143821               (not set)   
3016741                    NaN          anonymized                     NaN   
3016742                    NaN  cliqz.com/tracking                     NaN   

             brand  price  quantity device_category device_browser device_os  \
3016738  TT_Casual  36.97       1.0          mobile         Chrome   Android   
3016739  TT_Casual  42.85       2.0          mobile         Safari       iOS   
3016740  TT_Casual  37.39       2.0          mobile         Safari       iOS   
3016741        NaN    NaN       NaN             NaN            NaN       NaN   
3016742        NaN    NaN       NaN             NaN            NaN       NaN   

         country                  region  city_id          product_line  \
3016738  Germany                  Berlin  1003854         Home & Living   
3016739  Germany  North Rhine-Westphalia  1004707         Home & Living   
3016740  Germany                  Berlin  1003854         Home & Living   
3016741      NaN                     NaN      NaN  Shirts unter 20 Euro   
3016742      NaN                     NaN      NaN                 Damen   

                product_group     product_line_r           product_group_r  \
3016738              Vorhänge      Home & Living      Schuhe & Accessoires   
3016739              Vorhänge      Home & Living      Schuhe & Accessoires   
3016740              Vorhänge      Home & Living      Schuhe & Accessoires   
3016741  Shirts unter 20 Euro  Shirts unter 20 €  T-Shirts & Langarmshirts   
3016742               langarm              Damen  T-Shirts & Langarmshirts   

        campaign_involved  
3016738              True  
3016739              True  
3016740              True  
3016741               NaN  
3016742               NaN

In [13]:
df.shape

(3016743, 21)

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3016743 entries, 0 to 3016742
Data columns (total 21 columns):
visit_id_n           int64
unser_id_n           int64
visit_number         float64
date                 float64
article_name         object
article_id           object
ga_category          object
brand                object
price                float64
quantity             float64
device_category      object
device_browser       object
device_os            object
country              object
region               object
city_id              object
product_line         object
product_group        object
product_line_r       object
product_group_r      object
campaign_involved    object
dtypes: float64(4), int64(2), object(15)
memory usage: 483.3+ MB


In [15]:
df.describe()

visit_id_n    unser_id_n  visit_number          date         price  \
count  3.016743e+06  3.016743e+06  3.016732e+06  3.016732e+06  3.016732e+06   
mean   4.102715e+05  3.108908e+05  1.058004e+01  2.018055e+07  2.668266e+01   
std    2.366932e+05  1.795393e+05  2.962561e+01  4.429808e+03  1.790388e+01   
min    1.000000e+00  1.000000e+00  1.000000e+00  2.017110e+07  0.000000e+00   
25%    2.053810e+05  1.551850e+05  1.000000e+00  2.018032e+07  1.445000e+01   
50%    4.107240e+05  3.107390e+05  3.000000e+00  2.018071e+07  2.420000e+01   
75%    6.152210e+05  4.659450e+05  9.000000e+00  2.018103e+07  3.361000e+01   
max    8.201950e+05  6.226180e+05  1.616000e+03  2.019023e+07  2.512600e+02   

           quantity  
count  3.016732e+06  
mean   1.018650e+00  
std    1.586700e-01  
min    1.000000e+00  
25%    1.000000e+00  
50%    1.000000e+00  
75%    1.000000e+00  
max    1.000000e+01

In [16]:
df.sample(n=15)

visit_id_n  unser_id_n  visit_number        date  \
785418       180148      391682           1.0  20171116.0   
2446982      277567      114124           4.0  20180623.0   
1983080      676600      293503           4.0  20180719.0   
968955       220926      166167          20.0  20180319.0   
2835668      495432      276222           2.0  20181117.0   
2701234      223749      473927           1.0  20181030.0   
626523        63396      216891          18.0  20171124.0   
1115323      570681      307201           1.0  20181004.0   
1267382       45473       49109          66.0  20180516.0   
1059779      290073       62631          23.0  20190122.0   
1744021      454467      566487           5.0  20180406.0   
1400484      701657      355312           2.0  20180914.0   
1373484      453768      602537           4.0  20180716.0   
637326       599081      309555           3.0  20171101.0   
2227451      758783      224226           2.0  20181017.0   

                                  article_name     article_id  \
785418                      Alexa Skinny Jeans  4059953530467   
2446982     T-Shirt in Melange-Optik mit Print  4062105082301   
1983080              Bluse mit floralem Muster  4060868270560   
968955   Sweatshirt in Melange-Optik mit Print  4060463011124   
2835668                      Hemd mit Struktur  4062105675886   
2701234                 T-Shirt mit Logo-Print  4062105515564   
626523         Sweatshirt mit Schleifen-Detail  4059953136935   
1115323                    gemustertes T-Shirt  4060586088928   
1267382                 T-Shirt im Materialmix  4060586215300   
1059779                    T-Shirt mit Artwork  4060463174805   
1744021           Shirt mit Details aus Spitze  4060586866571   
1400484                  Marvin Straight Jeans  4060586342686   
1373484               Alexa Skinny Capri Jeans  4060586274444   
637326                          kariertes Hemd  4059953172797   
2227451              Cord-Hose mit Knopfleiste  4060868697374   

                                               ga_category      brand  price  \
785418                        Damen/Damenmode/Jeans/Skinny  TT_Casual  40.34   
2446982                                Herren/Print Shirts  TT_Casual   9.40   
1983080                    Damen/Damenmode/Blusen/ärmellos   TT_Denim  24.99   
968955   Kinder/Kindermode/Baby Mädchen 0 - 17 Monate/S...  (not set)  10.74   
2835668                                             Herren  TT_Casual  25.20   
2701234  Herren/Herrenmode/T-Shirts & Langarmshirts/T-S...  TT_Casual   7.56   
626523   Damen/Damenmode/Pullover & Sweatshirts/Sweatsh...   TT_Denim  12.60   
1115323                                    Baumwoll-Shirts  TT_Casual   5.04   
1267382  Damen/Damenmode/T-Shirts & Langarmshirts/T-Shirts  TT_Casual  20.16   
1059779  Kinder/Kindermode/Mädchen 1,5 - 7  Jahre/T-Shi...  TT_Casual   4.12   
1744021  Damen/Damenmode/T-Shirts & Langarmshirts/T-Shirts  (not set)  17.47   
1400484                    Herren/Herrenmode/Jeans/Regular  TT_Casual  42.01   
1373484                                  Damen/Neu im Sale  TT_Casual  29.40   
637326                    Herren/Herrenmode/Hemden/langarm   TT_Denim  33.61   
2227451                                              Hosen  TT_Casual  58.82   

         quantity device_category    device_browser  device_os  country  \
785418        1.0         desktop            Safari  Macintosh  Germany   
2446982       1.0         desktop             Opera    Windows  Germany   
1983080       1.0         desktop           Firefox    Windows  Austria   
968955        1.0          mobile  Samsung Internet    Android  Germany   
2835668       1.0         desktop              Edge    Windows  Germany   
2701234       1.0         desktop           Firefox    Windows  Germany   
626523        1.0          mobile            Chrome    Android  Germany   
1115323       1.0          mobile  Samsung Internet    Android  Germany   
1267382       1.0         desktop  

<a id='2.1'></a>

## Feature dictionary ( my impressions )

| Feature No. | Feature Name | Description |
| :-: | :-: | :-: |
| 1 | visit_id_n | Unique ID of a single shopping cart |
| 2 | unser_id_n | Unique ID of a user (misstyped obviously) |
| 3 | visit_number | Not entirely clear; either number of visits to the shop or number of total purchases |
| 4 | date | date the purchase was made |
| 5 | article_name | Name of the article that was bought |
| 6 | article_id | ID of the article that was bought |
| 7 | ga_category | Category the article falls in ( partly the path in the webshop to the article ) |
| 8 | brand | Brand of the article |
| 9 | price | Price of the article |
| 10 | quantity | Number of the article that were bought |
| 11 | device_category | Category the user device falls in |
| 12 | device_browser | Category the user browser falls in |
| 13 | device_os | Category the user OS falls in |
| 14 | country | country location of the user |
| 15 | region | region location of the user |
| 16 | city_id | geoID of the town location of the user |
| 17 | product_line | Product line of the article |
| 18 | product_group | Product group of the article |
| 19 | product_line_r | not clear what the r stands for; might stand for retail |
| 20 | product_group_r | not clear what the r stands for; might stand for retail |
| 21 | campaign_involved | not clear what the campaign stands for |

<a id='3'></a>

# 3. Data cleaning

In [17]:
#rename columns
df.rename(columns={'visit_id_n':'visit_id',
                          'unser_id_n':'user_id',
                          'ga_category':'category'}, inplace=True)

<a id='3.1'></a>

## 3.1 Correcting Visit ID
After extensive research in a later part of the project it came up, that some visit id's were assigned to multiple users. Probably because the german and austrian webshop data were just put together and not filtered. As such define visit id as the sum of visit id and user id to get a unique number.

In [18]:
df["visit_id"] = df["visit_id"].apply(lambda x: str(x)) + df["user_id"].apply(lambda x: str(x))
df = df.astype({"visit_id": "int64"})

<a id='3.2'></a>

## 3.2 Correcting Product group and line

After a talk with the people responsible for the data they concluded that product_line_r and product_group_r are the correct features, hence the ones without _r will be dropped.

In [19]:
df = df.drop(columns=["product_line", "product_group"])
df.rename(columns={"product_line_r": "product_line", "product_group_r": "product_group"}, inplace=True)

<a id='3.3'></a>

## 3.3 NaNs

In [20]:
df.isna().sum()

visit_id              0
user_id               0
visit_number         11
date                 11
article_name         11
article_id            0
category             11
brand                11
price                11
quantity             11
device_category      11
device_browser       11
device_os            11
country              11
region               11
city_id              11
product_line          0
product_group         0
campaign_involved    11
dtype: int64

In [21]:
df.dropna(inplace=True) #just 11 , so drop them

In [22]:
df.isna().sum()

visit_id             0
user_id              0
visit_number         0
date                 0
article_name         0
article_id           0
category             0
brand                0
price                0
quantity             0
device_category      0
device_browser       0
device_os            0
country              0
region               0
city_id              0
product_line         0
product_group        0
campaign_involved    0
dtype: int64

In [23]:
#number of (not set)
df.isin(['(not set)']).sum(axis=0)

visit_id                  0
user_id                   0
visit_number              0
date                      0
article_name              0
article_id                0
category              41935
brand                398925
price                     0
quantity                  0
device_category           0
device_browser            0
device_os                75
country                1178
region                 7731
city_id               93307
product_line          31227
product_group         31227
campaign_involved         0
dtype: int64

<a id='3.4'></a>

## 3.4 Datatypes

In [24]:
#change visit_number and quantity to int as float does not make sense
df = df.astype({'visit_number': 'int64', "quantity": "int64", "article_id": "int64"})
#device stuff and brand to category
df = df.astype({'device_category': 'category', "device_browser": "category", "device_os": "category", "brand": "category"})
#campaign_involved to bool
df.campaign_involved = df.campaign_involved.astype("bool")

In [25]:
#change objects to string for starters
#df[df.select_dtypes('object').columns] = df[df.select_dtypes('object').columns].astype(np.str) 
#does not work

In [26]:
#date to date format
df["date"]= pd.to_datetime(df["date"],format='%Y%m%d')

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3016732 entries, 0 to 3016740
Data columns (total 19 columns):
visit_id             int64
user_id              int64
visit_number         int64
date                 datetime64[ns]
article_name         object
article_id           int64
category             object
brand                category
price                float64
quantity             int64
device_category      category
device_browser       category
device_os            category
country              object
region               object
city_id              object
product_line         object
product_group        object
campaign_involved    bool
dtypes: bool(1), category(4), datetime64[ns](1), float64(1), int64(5), object(7)
memory usage: 359.6+ MB


### Price seems wrong for the following, as the original price is around 5-8, so discounts are in the price already

In [28]:
df.iloc[658646]

visit_id                                                  424670120783
user_id                                                         120783
visit_number                                                         4
date                                               2018-07-29 00:00:00
article_name                                  Bandana in Melange-Optik
article_id                                               4059953239858
category             Kinder/Kindermode/Baby Mädchen 0 - 17 Monate/A...
brand                                                        TT_Casual
price                                                             0.38
quantity                                                             1
device_category                                                desktop
device_browser                                                 Firefox
device_os                                                      Windows
country                                                        Germany
region

### Renaming a particular article

In [29]:
d = {'Geschenk Gutschein 10 EUR':'Geschenkgutschein 10 EUR','Geschenk Gutschein 20 EUR':'Geschenkgutschein 20 EUR','Geschenk Gutschein 50 EUR':'Geschenkgutschein 50 EUR'}
df = df.replace(d)

<a id='3.5'></a>

## 3.5 Duplicates

### Duplicated rows

In [30]:
df.duplicated().sum()

632

In [31]:
duplicateRowsDF = df[df.duplicated(keep=False)]

In [32]:
duplicateRowsDF.head()

visit_id  user_id  visit_number       date  \
118    22892413921   413921            45 2018-07-11   
119    22892413921   413921            45 2018-07-11   
675   177945413228   413228             1 2019-01-01   
676   177945413228   413228             1 2019-01-01   
1010  601457557264   557264             1 2018-12-25   

                      article_name     article_id  \
118       Geschenkgutschein 20 EUR   200000000020   
119       Geschenkgutschein 20 EUR   200000000020   
675   Sneaker-Socken im Doppelpack  4002702123007   
676   Sneaker-Socken im Doppelpack  4002702123007   
1010  Sneaker-Socken im Doppelpack  4002702123014   

                                               category      brand  price  \
118                                              Herren  (not set)  20.00   
119                                              Herren  (not set)  20.00   
675   Kinder/Kindermode/Mädchen 1,5 - 7  Jahre/Schuh...  TT_Casual   2.31   
676   Kinder/Kindermode/Mädchen 1,5 - 7  Jahre/Schuh...  TT_Casual   2.31   
1010  Kinder/Kindermode/Mädchen 1,5 - 7  Jahre/Schuh...  TT_Casual   3.18   

      quantity device_category device_browser device_os  country  \
118          1         desktop         Chrome   Windows  Austria   
119          1         desktop         Chrome   Windows  Austria   
675          1          mobile         Chrome   Android  Germany   
676          1          mobile         Chrome   Android  Germany   
1010        10          mobile         Chrome       iOS  Germany   

                 region  city_id product_line         product_group  \
118       Upper Austria  1000827       Herren                Herren   
119       Upper Austria  1000827       Herren                Herren   
675              Berlin  1003854       Kinder  Schuhe & Accessoires   
676              Berlin  1003854       Kinder  Schuhe & Accessoires   
1010  Baden-Wurttemberg  1004022       Kinder  Schuhe & Accessoires   

      campaign_involved  
118               False  
119               False  
675                True  
676                True  
1010              False

Note that it is not sure whether these rows are truly duplicates or just multiple purchases where the quantity wasn't set properly or the articles have the same name and price and for example different colors which this data does not tell us.

In [33]:
#df.drop_duplicates(keep="first", inplace=True)
#df.reset_index()

checking number of visits to number of entries (45 for this user)

In [34]:
df.user_id[413921]

230677

In [35]:
df[df['user_id'] == 413921].shape

(27, 19)

### Duplicates in visit_id

In [36]:
df.visit_id.max()

820195135969

In [37]:
df.visit_id.duplicated().sum()

2181687

In [38]:
df_visits = df[df.duplicated(subset="visit_id") == True]

In [39]:
df_uniquevisits = df.drop_duplicates(subset="visit_id")

In [40]:
df_visits.shape

(2181687, 19)

In [41]:
df_uniquevisits.shape

(835045, 19)

In [42]:
df_visits.sort_values(by=['visit_id']).head()

visit_id  user_id  visit_number       date  \
2972494    234440       40             1 2018-11-18   
2817928    234440       40             1 2018-11-18   
2538377    234440       40             1 2018-11-18   
2375327    276103    76103            22 2019-02-21   
2647054    290200      200            43 2018-11-03   

                         article_name     article_id  \
2972494              Schlichter Snood  4062105772851   
2817928              Hoodie mit Print  4062105670652   
2538377  Basic Herren-Strick-Pullover  4062105188904   
2375327          Culver Skinny Jeans   4061945274082   
2647054                    Steppjacke  4062105383903   

                               category      brand  price  quantity  \
2972494                  Hoodies/Herren   TT_Denim  17.99         1   
2817928                  Hoodies/Herren  TT_Casual  15.00         1   
2538377                          Herren  TT_Casual  14.99         1   
2375327  Herren/Herrenmode/Jeans/Skinny   TT_Denim  42.01         1   
2647054                     Steppjacken  TT_Casual  31.51         1   

        device_category device_browser device_os  country         region  \
2972494          mobile         Chrome   Android  Austria  Lower Austria   
2817928          mobile         Chrome   Android  Austria  Lower Austria   
2538377          mobile         Chrome   Android  Austria  Lower Austria   
2375327          mobile         Safari       iOS  Germany         Berlin   
2647054          mobile         Chrome   Android  Germany    Brandenburg   

         city_id    product_line           product_group  campaign_involved  
2972494  1000750          Herren  Pullover & Sweatshirts               True  
2817928  1000750          Herren                  Herren               True  
2538377  1000750          Herren  Pullover & Sweatshirts               True  
2375327  1003854          Herren                   Jeans               True  
2647054  1003866  Herbst Outfits          Herbst Outfits               True

In [43]:
df_uniquevisits.sort_values(by=["visit_id"]).head()

visit_id  user_id  visit_number       date  \
320852     234440       40             1 2018-11-18   
2079241    276103    76103            22 2019-02-21   
2645748    290200      200            43 2018-11-03   
2830805    336120       20             2 2018-10-25   
321178     366022     6022             8 2017-12-30   

                         article_name     article_id  \
320852   Basic Herren-Strick-Pullover  4059491520210   
2079241               Henley T-Shirt   4060868345305   
2645748                    Steppjacke  4062105383569   
2830805       Nela Extra Skinny Jeans  4062105674247   
321178   Basic Herren-Strick-Pullover  4059491520241   

                                                  category      brand  price  \
320852                                              Herren  TT_Casual  14.99   
2079241  Herren/Herrenmode/T-Shirts & Langarmshirts/T-S...  TT_Casual  10.91   
2645748                                  Steppjacken/Damen  TT_Casual  31.50   
2830805                       Damen/Damenmode/Jeans/Skinny   TT_Denim  33.61   
321178                                    Feeling Blue Men  TT_Casual  16.80   

         quantity device_category device_browser  device_os  country  \
320852          1          mobile         Chrome    Android  Austria   
2079241         1          mobile         Safari        iOS  Germany   
2645748         1          mobile         Chrome    Android  Germany   
2830805         1          mobile         Chrome    Android  Germany   
321178          1         desktop         Safari  Macintosh  Germany   

                region  city_id      product_line             product_group  \
320852   Lower Austria  1000750            Herren    Pullover & Sweatshirts   
2079241         Berlin  1003854            Herren  T-Shirts & Langarmshirts   
2645748    Brandenburg  1003866             Damen   Jacken, Mäntel & Westen   
2830805       Saarland  9048444             Damen                     Jeans   
321178          Berlin  1003854  Feeling Blue Men          Feeling Blue Men   

         campaign_involved  
320852                True  
2079241               True  
2645748               True  
2830805               True  
321178                True

Duplicates in the visit_id surely denote multiple articles that were bought at the same visit, thus dropping any will not make sense. Also there were 820194 unique visit_id's, hence we know that the data consists of this number of shopping cart purchases.

<a id='4'></a>

# 4. Understanding particular Features in Detail

## Article names and id

In [44]:
df.article_name.value_counts().head(10)

Alexa Slim Jeans                          62458
Josh Regular Slim Jeans                   41092
T-Shirt mit Schrift-Print                 38895
T-Shirt mit Print                         29257
T-Shirt mit Logo-Print                    28231
Marvin Straight Jeans                     27405
Alexa Straight Jeans                      22816
Damen-Basic-Pullover mit V-Ausschnitt     19512
Jim Slim Bermuda Shorts                   18002
Basic Herren-Pullover mit V-Ausschnitt    17812
Name: article_name, dtype: int64

In [45]:
df.article_id.value_counts().head(10)

4059066698719    1728
4059066698511    1284
4002702235090    1215
4059066698726    1176
4059066698733    1054
4059953383155     998
4059953383148     908
4059953383162     887
4059953383179     861
4059491897060     819
Name: article_id, dtype: int64

## Unique devices 
lets check the entries and number of occurance of the elements of the device features

In [46]:
df.device_browser.value_counts()

Chrome                      994731
Safari                      793300
Firefox                     589700
Samsung Internet            256273
Edge                        155931
Internet Explorer            88370
Android Webview              63918
Opera                        29834
Safari (in-app)              28517
Amazon Silk                  14662
YaBrowser                      354
SeaMonkey                      337
Puffin                         235
Maxthon                        217
Android Browser                 94
Mozilla Compatible Agent        85
Iron                            58
UC Browser                      43
Playstation 4                   26
BlackBerry                      14
MRCHROME                        12
Seznam                          12
Coc Coc                          4
Opera 9.4                        3
Nintendo Browser                 2
Name: device_browser, dtype: int64

In [47]:
df.device_category.value_counts()

desktop    1371610
mobile     1091610
tablet      553512
Name: device_category, dtype: int64

In [48]:
df.device_os.value_counts()

Windows          1204420
Android           923303
iOS               718614
Macintosh         156569
Linux              11633
Chrome OS           1778
BlackBerry           222
(not set)             75
Windows Phone         70
Playstation 4         26
Tizen                 19
Nintendo WiiU          2
FreeBSD                1
Name: device_os, dtype: int64

Nothing too surprising so far. Later we can check if there were differences in the purchases between the different features.

## Campaign_involved:

In [49]:
camp_true = df[df.campaign_involved == True]

In [50]:
len(camp_true)

2278279

In [51]:
camp_true.sample(n=5)

visit_id  user_id  visit_number       date  \
211198    68710245653   245653            12 2018-06-23   
1548856  424622297659   297659            75 2018-08-22   
1651421   19840287065    87065             7 2018-03-20   
2778814  756228225980   225980            15 2019-01-16   
2147697  381538211367   211367             5 2018-10-17   

                              article_name     article_id  \
211198                Slipper mit Schleife  4058219402630   
1548856  Sweatshirt mit floraler Stickerei  4060586637874   
1651421               Tapered Relaxed Hose  4060586736782   
2778814        Pullover mit Strukturmuster  4062105573335   
2147697                  Gemustertes Shirt  4060868496137   

                                                  category      brand  price  \
211198                                              Schuhe   TT_Denim  26.63   
1548856  Damen/Damenmode/Pullover & Sweatshirts/Sweatsh...   TT_Denim  10.08   
1651421                              Damen/Damenmode/Hosen  TT_Casual  47.05   
2778814    Damen/Damenmode/Pullover & Sweatshirts/Pullover  TT_Casual  29.40   
2147697                                      Langarmshirts  TT_Casual  24.99   

         quantity device_category device_browser device_os  country  \
211198          1          mobile         Chrome   Android  Austria   
1548856         1          mobile         Chrome   Android  Germany   
1651421         1          mobile         Chrome   Android  Germany   
2778814         1          mobile         Chrome   Android  Germany   
2147697         1          mobile         Chrome   Android  Austria   

                         region  city_id          product_line  \
211198                   Vienna  1000997  Schuhe & Accessoires   
1548856  North Rhine-Westphalia  1004607                 Damen   
1651421                 Bavaria  1004242                 Damen   
2778814            Lower Saxony  1004537                 Damen   
2147697               Carinthia  1000700                 Damen   

                    product_group  campaign_involved  
211198       Schuhe & Accessoires               True  
1548856    Pullover & Sweatshirts               True  
1651421                     Hosen               True  
2778814    Pullover & Sweatshirts               True  
2147697  T-Shirts & Langarmshirts               True

In [52]:
camp_false = df[df.campaign_involved == False]

In [53]:
len(camp_false)

738453

In [54]:
camp_false[camp_false["article_id"]==4011694579266].head()

visit_id  user_id  visit_number       date           article_name  \
35271  505889355464   355464             5 2017-11-13  Hip-Pants im 3er-Pack   
35274  383063313091   313091             2 2018-03-24  Hip-Pants im 3er-Pack   
35277  201946434776   434776             6 2018-07-29  Hip-Pants im 3er-Pack   
35281  333769606504   606504             1 2018-09-17  Hip-Pants im 3er-Pack   
35295   14858564131   564131             1 2019-02-04  Hip-Pants im 3er-Pack   

          article_id                               category      brand  price  \
35271  4011694579266  Herren/Herrenmode/Underwear & Pyjamas  TT_Casual  15.12   
35274  4011694579266  Herren/Herrenmode/Underwear & Pyjamas  TT_Casual  18.90   
35277  4011694579266  Herren/Herrenmode/Underwear & Pyjamas  TT_Casual  25.20   
35281  4011694579266  Herren/Herrenmode/Underwear & Pyjamas  TT_Casual  20.16   
35295  4011694579266  Herren/Herrenmode/Underwear & Pyjamas  TT_Casual  16.80   

       quantity device_category device_browser  device_os  country  \
35271         1          mobile        Firefox    Android  Germany   
35274         1          tablet         Safari        iOS  Germany   
35277         1         desktop        Firefox    Windows  Germany   
35281         1         desktop         Chrome  Macintosh  Germany   
35295         1         desktop        Firefox  Macintosh  Germany   

                   region  city_id product_line        product_group  \
35271              Berlin  1003854       Herren  Underwear & Pyjamas   
35274  Schleswig-Holstein  9048708       Herren  Underwear & Pyjamas   
35277  Schleswig-Holstein  1004866       Herren  Underwear & Pyjamas   
35281              Berlin  1003854       Herren  Underwear & Pyjamas   
35295        Lower Saxony  1004536       Herren  Underwear & Pyjamas   

       campaign_involved  
35271              False  
35274              False  
35277              False  
35281              False  
35295              False

In [55]:
camp_true[camp_true["article_id"]==4011694579266].head()

visit_id  user_id  visit_number       date           article_name  \
35270  723099348206   348206             6 2017-11-13  Hip-Pants im 3er-Pack   
35272  539377527591   527591             6 2018-08-29  Hip-Pants im 3er-Pack   
35273   54340111696   111696             2 2018-09-08  Hip-Pants im 3er-Pack   
35275  492320350403   350403             6 2018-10-26  Hip-Pants im 3er-Pack   
35276   31699418198   418198             1 2018-09-13  Hip-Pants im 3er-Pack   

          article_id                               category      brand  price  \
35270  4011694579266  Herren/Herrenmode/Underwear & Pyjamas  TT_Casual  21.42   
35272  4011694579266  Herren/Herrenmode/Underwear & Pyjamas  TT_Casual  25.21   
35273  4011694579266  Herren/Herrenmode/Underwear & Pyjamas  TT_Casual  18.90   
35275  4011694579266  Herren/Herrenmode/Underwear & Pyjamas  TT_Casual  14.70   
35276  4011694579266  Herren/Herrenmode/Underwear & Pyjamas  TT_Casual  18.90   

       quantity device_category    device_browser  device_os  country  \
35270         1          mobile  Samsung Internet    Android  Germany   
35272         1          mobile  Samsung Internet    Android  Germany   
35273         1         desktop            Chrome  Chrome OS  Germany   
35275         2         desktop            Chrome    Windows  Germany   
35276         1          mobile            Chrome    Android  Germany   

                       region  city_id product_line        product_group  \
35270                  Berlin  1003854       Herren  Underwear & Pyjamas   
35272                 Bavaria  1004269       Herren  Underwear & Pyjamas   
35273            Lower Saxony  9048874       Herren  Underwear & Pyjamas   
35275  North Rhine-Westphalia  1004761       Herren  Underwear & Pyjamas   
35276               Thuringia  1004999       Herren  Underwear & Pyjamas   

       campaign_involved  
35270               True  
35272               True  
35273               True  
35275               True  
35276               True

we see, that campaign_involved can not mean discounts

## Number of occurence of countries

In [56]:
df.country.value_counts().head()

Germany        2756836
Austria         196950
Netherlands      22547
Switzerland       5425
Belgium           2434
Name: country, dtype: int64

## Check for missing dates 

In [57]:
from datetime import date, timedelta
date_set = set(df.date.min() + timedelta(x) for x in range((df.date.max() - df.date.min()).days))
missing = sorted(date_set - set(df.date))
missing

[]

As we can see there are no missing days in the date Data.

<a id='5'></a>

# 5. Feature Engineering

<a id='5.1'></a>

## 5.1 Cities by Code

In [58]:
df_geotargets.head()

Criteria ID        Name                            Canonical Name  \
0      1000002       Kabul                   Kabul,Kabul,Afghanistan   
1      1000003      Luanda             Luanda,Luanda Province,Angola   
2      1000004  The Valley                       The Valley,Anguilla   
3      1000010   Abu Dhabi  Abu Dhabi,Abu Dhabi,United Arab Emirates   
4      1000011       Ajman          Ajman,Ajman,United Arab Emirates   

   Parent ID Country Code Target Type  Status  
0  9075393.0           AF        City  Active  
1  9070431.0           AO        City  Active  
2     2660.0           AI        City  Active  
3  9041082.0           AE        City  Active  
4  9047096.0           AE        City  Active

In [59]:
#Criteria ID as index
df_geotargets_clean = df_geotargets.set_index("Criteria ID")

In [60]:
#take just the Name column
df_geotargets_clean = df_geotargets_clean["Name"]

In [61]:
#get a dictionary from the dataframe
geodic = df_geotargets_clean.to_dict()

In [62]:
#entry for not set ids
geodic[0] = "(not set)"

In [63]:
#replace not set with a number for a moment to change to int
df["city_id"] = df["city_id"].replace({"(not set)": 0})

In [64]:
#change city_id type to int
df = df.astype({'city_id': 'int64'})

In [65]:
#map cities to dataframe with new column city
df["city"] = df["city_id"].map(geodic)

In [66]:
#nans in cities by ids that were not found
df.isna().sum()

visit_id               0
user_id                0
visit_number           0
date                   0
article_name           0
article_id             0
category               0
brand                  0
price                  0
quantity               0
device_category        0
device_browser         0
device_os              0
country                0
region                 0
city_id                0
product_line           0
product_group          0
campaign_involved      0
city                 640
dtype: int64

In [67]:
#df[df["city"].isnull()]

In [68]:
#fill the nans with (not set)
df.fillna(value={"city": "(not set)"}, inplace=True)

In [69]:
#df[df["city"].isnull()]["country"].value_counts()

In [70]:
df_geotargets["Target Type"].value_counts()

Postal Code               48391
City                      38097
Neighborhood               4631
County                     3417
Municipality               2124
Province                   1131
District                    948
Region                      909
Congressional District      441
Airport                     395
Department                  240
State                       235
University                  219
Country                     213
City Region                 183
Governorate                 121
National Park                96
Borough                      88
Prefecture                   49
Okrug                        28
Canton                       26
Autonomous Community         22
TV Region                    14
Union Territory               7
Territory                     4
Name: Target Type, dtype: int64

In [71]:
# check of the postal codes for germany ( not useful )
#df_geotargets[df_geotargets["Target Type"] == "Postal Code"][df_geotargets["Country Code"]=="DE"].head()

<a id='5.2'></a>

## 5.2 Timespan

In [72]:
year_list = pd.DatetimeIndex(df['date']).year  

In [73]:
year_list.unique()

Int64Index([2018, 2017, 2019], dtype='int64', name='date')

In [74]:
df.date.max()

Timestamp('2019-02-28 00:00:00')

In [75]:
df.date.min()

Timestamp('2017-11-01 00:00:00')

We have data from 2017-11-01 till 2019-02-28

<a id='5.3'></a>

## 5.3 Weekdays
monday 0 , ... , sunday 6

In [76]:
df["day_of_week"] = pd.to_datetime(df["date"]).dt.dayofweek

<a id='5.4'></a>

## 5.4 Article Genre

In [77]:
#put all features describing the article type together for easier use
df["genre_column_placeholder"] = df["article_name"] + "/" + df["category"] + "/" + df["product_line"] + "/" + df["product_group"]

In [78]:
df.genre_column_placeholder = df.genre_column_placeholder.astype("str")

In [79]:
#new column for articles of women genre
def check_women(item):
    return "" if re.search(r"([Dd]{1}ame[ns]{1})|([Ww]{1}om[ea]{1}n)", item) == None else "Women"

df['women_article'] = df['genre_column_placeholder'].apply(check_women)


In [80]:
#new column for articles of men genre
def check_men(item):
    return "" if re.search(r"([Hh]{1}e[r]{1,2}en)|(Men\b)|(\b[Mm]en\b)", item) == None else "Men"

df['men_article'] = df['genre_column_placeholder'].apply(check_men)


In [81]:
#new column for articles of kids/teenager/babies genre
def check_kids(item):
    return "" if re.search(r"([Kk]{1}inder)|(\b[Kk]{1}id[s]{0,1})|([Jj]{1}ungen)|([Mm]{1}ädchen)|([Tt]{1}een[s]{0,1})|([Bb]{1}aby)|([Bb]{1}abies)", item) == None else "Kids"

df['kids_article'] = df['genre_column_placeholder'].apply(check_kids)


In [82]:
#new column for articles of home&living genre
def check_hl(item):
    return "" if re.search(r"([Hh]{1}ome)|([Ll]iving)", item) == None else "Home&living"

df['homeandliving_article'] = df['genre_column_placeholder'].apply(check_hl)


In [83]:
df = df.assign(category_slim = df.women_article + df.men_article + df.kids_article + df.homeandliving_article)

In [84]:
def putslash(item): 
  
    # regex [A-Z][a-z]* means any string starting  
    # with capital character followed by many  
    # lowercase letters  
    words = re.findall('[A-Z][a-z]*', item) 
  
    # Change first letter of each word into lower 
    # case 
    result = [] 
    for word in words: 
        word = chr( ord (word[0]) + 32) + word[1:] 
        result.append(word) 
    return '/'.join(result)

In [85]:
df["category_slim"] = df["category_slim"].apply(putslash)

In [86]:
df["category_slim"] = df["category_slim"].replace("", "(not set)")

In [87]:
df.women_article = df.women_article.replace({"": 0, "Women": 1})
df.men_article = df.men_article.replace({"": 0, "Men": 1})
df.kids_article = df.kids_article.replace({"": 0, "Kids": 1})
df.homeandliving_article = df.homeandliving_article.replace({"": 0, "Home&living": 1})

In [88]:
article_diff = df[(df.women_article + df.men_article + df.kids_article + df.homeandliving_article) == 0]

In [89]:
article_diff.product_group.value_counts()

Schuhe & Accessoires        7885
(not set)                   7284
Herbst Outfits              6384
T-Shirts & Langarmshirts    1052
Neu im Sale                  427
Festival Styles              326
SAFARI STYLES                212
Summer Sale Highlights       121
Bold Summer Colors            83
Black & White                 46
Alles zum halben Preis        24
T-Shirts unter 20 €            9
Partyoutfits                   3
Revolverheld Kollektion        2
Mexico Styles                  2
Hosen                          1
Name: product_group, dtype: int64

In [90]:
article_diff.category.value_counts().head()

(not set)         7000
Herbst Outfits    4431
Athleisure        3881
Schuhe            1812
Winterpullover     818
Name: category, dtype: int64

In [91]:
len(article_diff)

23861

So we are left with only 23861 articles without any of the new article groups.

In [92]:
#number of articles that got multiple article groups
mult_article = df[df.women_article + df.men_article + df.kids_article + df.homeandliving_article > 1]

In [93]:
mult_article.sample(n=5)

visit_id  user_id  visit_number       date  \
2765945  777425313566   313566             1 2018-11-19   
2584675  255578243423   243423             1 2018-05-15   
266692   820041540406   540406             1 2018-07-14   
2743479   53058419601    19601             2 2018-10-28   
2340228  192234417025   417025             1 2018-11-16   

                                 article_name     article_id  \
2765945                      Parka mit Kapuze  4062105565316   
2584675                     Mia Slim 7/8 Hose  4062105311845   
266692         Skinny Herren-Chino mit Gürtel  4059066379984   
2743479               Gefütterter Steppmantel  4062105536705   
2340228  Alexa Slim Jeans mit seitlichem Tape  4061945170261   

                      category      brand   price  quantity device_category  \
2765945      Wintermode/Herren  TT_Casual  113.44         1         desktop   
2584675  Damen/Damenmode/Hosen  (not set)   33.61         1         desktop   
266692                   Hosen   TT_Denim   25.87         1         desktop   
2743479      Steppjacken/Damen  TT_Casual  135.98         1          mobile   
2340228       denimflash/Damen  TT_Casual   42.01         1          mobile   

           device_browser device_os  country                  region  city_id  \
2765945              Edge   Windows  Germany                 Bavaria  1004204   
2584675            Chrome   Windows  Germany  North Rhine-Westphalia  1004761   
266692             Chrome   Windows  Germany       Baden-Wurttemberg  1004054   
2743479  Samsung Internet   Android  Germany               Thuringia  9048414   
2340228            Safari       iOS  Germany  North Rhine-Westphalia  1004611   

        product_line product_group  campaign_involved                   city  \
2765945        Damen        Herren               True              Karlsfeld   
2584675       Herren         Hosen               True         Wermelskirchen   
266692         Damen         Hosen              False              Stuttgart   
2743479       Herren         Damen               True  Heilbad Heiligenstadt   
2340228       Herren         Damen               True               Dortmund   

         day_of_week                           genre_column_placeholder  \
2765945            0    Parka mit Kapuze/Wintermode/Herren/Damen/Herren   
2584675            1  Mia Slim 7/8 Hose/Damen/Damenmode/Hosen/Herren...   
266692             5   Skinny Herren-Chino mit Gürtel/Hosen/Damen/Hosen   
2743479            6  Gefütterter Steppmantel/Steppjacken/Damen/Herr...   
2340228            4  Alexa Slim Jeans mit seitlichem Tape/denimflas...   

         women_article  men_article  kids_article  homeandliving_article  \
2765945              1            1             0                      0   
2584675              1            1             0                      0   
266692               1            1             0                      0   
2743479              1            1             0                      0   
2340228              1            1             0                      0   

        category_slim  
2765945     women/men  
2584675     women/men  
266692      women/men  
2743479     women/men  
2340228     women/men

In [94]:
len(mult_article)

65808

Having 65808 articles with multiple article groups. Most of them do make at least some sense, like a jacket for a teenager also having the product_line Men.

<a id='5.5'></a>

## 5.5 Total price per article purchase

In [95]:
df["total_price"] = df["price"] * df["quantity"]

## Ordering columns

In [96]:
#code for checking the index of a certain column
df.columns.get_loc("price")

8

In [97]:
#check name of certain index
df.columns[27]

'total_price'

In [98]:
#reordering the columns
cols = df.columns.tolist()
cols = cols[0:4] + [cols[20]] + cols[4:6] + cols[8:10] + [cols[27]] \
        + cols[13:15] + [cols[19]] + [cols[15]] + cols[22:26] + [cols[18]] \
        + [cols[6]] + [cols[26]] + cols[7:8] + cols[16:18] + [cols[21]] + cols[10:13]
print("Check if any column was skipped when reordering (True: all columns fetched):", len(cols) == len(df.columns))
df = df.reindex(columns = cols)

Check if any column was skipped when reordering (True: all columns fetched): True


<a id='5.6'></a>

## 5.6 Value and number of items of shopping carts

In [99]:
shopcart = df.groupby(["visit_id"]).sum()

In [100]:
#take just the Name column
df_shopcartvalue_clean = shopcart["total_price"]
df_shopcartnumber_clean = shopcart["quantity"]

In [101]:
#get a dictionary from the dataframe
cartdicval = df_shopcartvalue_clean.to_dict()
cartdicnum = df_shopcartnumber_clean.to_dict()
cartdiffart = df.visit_id.value_counts().to_dict()

In [102]:
df["shop_cart_value"] = df["visit_id"].map(cartdicval)
df["shop_cart_item_count"] = df["visit_id"].map(cartdicnum)
df["shop_cart_unique_articles"] = df["visit_id"].map(cartdiffart)

In [103]:
#reordering the columns
cols = df.columns.tolist()
cols = cols[0:10] + cols[28:31] + cols[10:28]
print("Check if any column was skipped when reordering (True: all columns fetched):", len(cols) == len(df.columns))
df = df.reindex(columns = cols)

Check if any column was skipped when reordering (True: all columns fetched): True


<a id='5.7'></a>

## 5.7 Stockprice per day

In [104]:
df_stock["timestamp"]= pd.to_datetime(df_stock["timestamp"],format='%Y-%m-%d')

In [105]:
#fill missing days ( weekends )
df_stock = df_stock.resample('D',on="timestamp").mean()

In [106]:
#fill the nans of the missings above with the average of their direct set neighbours
df_stock = (df_stock.ffill() + df_stock.bfill())/2

In [107]:
df_stock.head()

open      high      low    close     volume
timestamp                                                 
2010-03-26  11.9637  12.04200  11.4576  11.8257  2767746.0
2010-03-27  11.8740  11.97755  11.5404  11.8349  1489778.5
2010-03-28  11.8740  11.97755  11.5404  11.8349  1489778.5
2010-03-29  11.7843  11.91310  11.6232  11.8441   211811.0
2010-03-30  11.8073  11.91770  11.6600  11.8717    80096.0

In [108]:
#take just the close column
df_stock_clean = df_stock["close"]

In [109]:
#get a dictionary from the dataframe
stockdic = df_stock_clean.to_dict()

In [110]:
#map stock to dataframe with new column stock_value
df["stock_value"] = df["date"].map(stockdic)

In [111]:
#check for nans that were not set 
df.isna().sum()

visit_id                     0
user_id                      0
visit_number                 0
date                         0
day_of_week                  0
article_name                 0
article_id                   0
price                        0
quantity                     0
total_price                  0
shop_cart_value              0
shop_cart_item_count         0
shop_cart_unique_articles    0
country                      0
region                       0
city                         0
city_id                      0
women_article                0
men_article                  0
kids_article                 0
homeandliving_article        0
campaign_involved            0
category                     0
category_slim                0
brand                        0
product_line                 0
product_group                0
genre_column_placeholder     0
device_category              0
device_browser               0
device_os                    0
stock_value                  0
dtype: i

In [112]:
#easy dataframe to check the stockprice development 
df.drop_duplicates("date").sort_values("date").head(5)

visit_id  user_id  visit_number       date  day_of_week  \
2325   14930374362   374362             3 2017-11-01            2   
5836  645245525600   525600             2 2017-11-02            3   
2849  448437326618   326618             1 2017-11-03            4   
2512  813967575820   575820            11 2017-11-04            5   
455   754081156504   156504             1 2017-11-05            6   

                     article_name     article_id  price  quantity  \
2325  Tennis-Socken im Dreierpack  4002702234727   5.01         1   
5836   Basic Socken im Doppelpack  4002702235090   5.84         2   
2849  Tennis-Socken im Dreierpack  4002702234765   5.52         2   
2512  Tennis-Socken im Dreierpack  4002702234758   6.68         2   
455      Geschenkgutschein 50 EUR   200000000050  50.00         1   

      total_price  shop_cart_value  shop_cart_item_count  \
2325         5.01            32.69                     5   
5836        11.68           108.55                     5   
2849        11.04            83.18                     7   
2512        13.36            78.38                     4   
455         50.00            50.00                     1   

      shop_cart_unique_articles  country                  region  \
2325                          5  Germany  North Rhine-Westphalia   
5836                          4  Germany       Baden-Wurttemberg   
2849                          6  Germany  North Rhine-Westphalia   
2512                          3  Germany       Baden-Wurttemberg   
455                           1  Germany      Schleswig-Holstein   

                city  city_id  women_article  men_article  kids_article  \
2325           Neuss  1004708              0            1             0   
5836       (not set)        0              0            1             0   
2849  Wermelskirchen  1004761              0            1             0   
2512       Esslingen  1003938              0            1             0   
455        Flensburg  1004855              0            1             0   

      homeandliving_article  campaign_involved  \
2325                      0               True   
5836                      0               True   
2849                      0              False   
2512                      0               True   
455                       0              False   

                                               category category_slim  \
2325  Herren/Herrenschuhe & Accessoires/Accessoires/...           men   
5836  Herren/Herrenschuhe & Accessoires/Accessoires/...           men   
2849  Herren/Herrenschuhe & Accessoires/Accessoires/...           men   
2512  Herren/Herrenschuhe & Accessoires/Accessoires/...           men   
455                                              Herren           men   

          brand product_line         product_group  \
2325  TT_Casual       Herren  Schuhe & Accessoires   
5836  TT_Casual       Herren  Schuhe & Accessoires   
2849  TT_Casual       Herren  Schuhe & Accessoires   
2512  TT_Casual       Herren  Schuhe & Accessoires   
455   (not set)       Herren                Herren   

                               genre_column_placeholder device_category  \
2325  Tennis-Socken im Dreierpack/Herren/Herrenschuh...          mobile   
5836  Basic Socken im Doppelpack/Herren/Herrenschuhe...         desktop   
2849  Tennis-Socken im Dreierpack/Herren/Herrenschuh...         desktop   
2512  Tennis-Socken im Dreierpack/Herren/Herrenschuh...          mobile   
455       Geschenkgutschein 50 EUR/Herren/Herren/Herren         desktop   

        device_browser device_os  stock_value  
2325            Chrome   Android        8.859  
5836           Firefox   Windows        9.131  
2849           Firefox   Windows        9.100  
2512  Samsung Internet   Android        9.090  
455            Firefox   Windows        9.090

<a id='5.8'></a>

## 5.8 First date of purchase and purchase type

In [113]:
#help feature with a edited date
df['year_month'] = df['date'].map(lambda date: 100*date.year + date.month)

In [114]:
#finding the date where users bought for the first time in our timeline
min_purchase = df.groupby("user_id").year_month.min().reset_index()
min_purchase.columns = ["user_id","min_purchase_date"]

#merge
df = pd.merge(df, min_purchase, on="user_id")

In [115]:
#create a feature user_type with first or established depending on time of purchase
df["user_type"] = "first"
df.loc[df["year_month"]>df["min_purchase_date"], "user_type"] = "established"

<a id='5.9'></a>

## 5.9 Final Feature dictionary

| Feature No. | Feature Name | Description |
| :-: | :-: | :-: |
| 1 | visit_id | Unique ID consisting of the ID of a single shopping cart and the ID of the user |
| 2 | user_id | Unique ID of a user |
| 3 | visit_number | Not entirely clear; either number of visits to the shop or number of total purchases |
| 4 | date | Date the purchase was made |
| 5 | day_of_week | Day when the purchase was made; 0 for monday till 6 for sunday |
| 6 | article_name | Name of the article that was bought |
| 7 | article_id | ID of the article that was bought |
| 8 | price | Price of the article |
| 9 | quantity | Number of the article that were bought |
| 10 | total_price | Total price of the article purchase consisting of price * quantity |
| 11 | shop_cart_value | Value of the whole shopping cart of a single visit_id |
| 12 | shop_cart_item_count | Number of items in the shopping cart of a single visit_id |
| 13 | shop_cart_unique_articles | Number of unique items in the shopping cart of a single visit_id |
| 14 | country | country location of the user |
| 15 | region | region location of the user |
| 16 | city | city location of the user |
| 17 | city_id | geoID of the town location of the user |
| 18 | women_article | women category - 1 yes , 0 no |
| 19 | men_article | men category - 1 yes , 0 no |
| 20 | kids_article | kids category - 1 yes , 0 no |
| 21 | homeandliving_article | home & living category - 1 yes , 0 no |
| 22 | campaign_involved | still not completely clear but after some research it might stand for the cases where users came to the shop via advertisement. When you get to the shop this way the url contains a part for campaign: for example "utm_campaign=de_retargeting_criteo"|
| 23 | category | Category the article falls in ( partly the path in the webshop to the article ) |
| 24 | category_slim | featured category with women, men, kids and home&living |
| 25 | brand | Brand of the article |
| 26 | product_line | Product line of the article |
| 27 | product_group | Product group of the article |
| 28 | genre_column_placeholder | Sum of article name, category and product lines/groups to better work with |
| 29 | device_category | Category the user device falls in |
| 30 | device_browser | Category the user browser falls in |
| 31 | device_os | Category the user OS falls in |
| 32 | stock_value | company stock price at this day |
| 33 | year_month | feature with year and month as int |
| 34 | min_purchase_date | date of the first purchase in our timeline of this user |
| 35 | user_type | type of the purchase, first or established |

## save dataframe to pickle

In [116]:
df.to_pickle('Data/dataframefull.pickle')